Index: cache.php
===================================================================
--- cache.php	(revision 0)
+++ cache.php	(revision 0)
@@ -0,0 +1,10 @@
+<?php 
+
+$lang = array
+(
+	'undefined_group'      => 'El grupo %s no esta definido en la configuracion.',
+	'extension_not_loaded' => 'La extensión PHP %s tiene que estar cargada para poder utilizar este driver.',
+	'unwritable'           => 'El directorio seleccionado, %s, no tiene permisos de escritura.',
+	'resources'            => 'No es posible guardar el contenido en la cache, el contenido no es serializable.',
+	'driver_error'         => '%s',
+);
\ No newline at end of file
Index: calendar.php
===================================================================
--- calendar.php	(revision 3284)
+++ calendar.php	(working copy)
@@ -10,7 +10,6 @@
 	'th'        => 'Ju',
 	'fr'        => 'Vi',
 	'sa'        => 'Sa',
-
 	// Short day names
 	'sun'       => 'Dom',
 	'mon'       => 'Lun',
@@ -19,16 +18,14 @@
 	'thu'       => 'Jue',
 	'fri'       => 'Vie',
 	'sat'       => 'Sab',
-
 	// Long day names
-	'sunday'    => 'Dominfo',
+	'sunday'    => 'Domingo',
 	'monday'    => 'Lunes',
 	'tuesday'   => 'Martes',
 	'wednesday' => 'Miercoles',
 	'thursday'  => 'Jueves',
 	'friday'    => 'Viernes',
 	'saturday'  => 'Sabado',
-
 	// Short month names
 	'jan'       => 'Ene',
 	'feb'       => 'Feb',
@@ -42,7 +39,6 @@
 	'oct'       => 'Oct',
 	'nov'       => 'Nov',
 	'dec'       => 'Dic',
-
 	// Long month names
 	'january'   => 'Enero',
 	'february'  => 'Febrero',
Index: captcha.php
===================================================================
--- captcha.php	(revision 0)
+++ captcha.php	(revision 0)
@@ -0,0 +1,33 @@
+<?php
+
+$lang = array
+(
+	'file_not_found'	=> 'El archivo especificado, %s, no ha sido encontrado. Por favor, verifica que el fichero existe utilizando file_exists() antes de intentar utilizarlo.',
+	'requires_GD2'		=> 'La libreria Captcha requiere GD2 con soporte FreeType. Lea lo siguiente http://php.net/gd_info para ampliar la informacion.',
+	
+	// Words of varying length for the Captcha_Word_Driver to pick from
+	// Note: use only alphanumeric characters
+	'words' => array
+	(
+		'cd', 'tv', 'it', 'to', 'be', 'or',
+		'sun', 'car', 'dog', 'bed', 'kid', 'egg',
+		'bike', 'tree', 'bath', 'roof', 'road', 'hair',
+		'hello', 'world', 'earth', 'beard', 'chess', 'water',
+		'barber', 'bakery', 'banana', 'market', 'purple', 'writer',
+		'america', 'release', 'playing', 'working', 'foreign', 'general',
+		'aircraft', 'computer', 'laughter', 'alphabet', 'kangaroo', 'spelling',
+		'architect', 'president', 'cockroach', 'encounter', 'terrorism', 'cylinders',
+	),
+
+	// Riddles for the Captcha_Riddle_Driver to pick from
+	// Note: use only alphanumeric characters
+	'riddles' => array
+	(
+		array('¿Odias el spam? (si o no)', 'si'),
+		array('¿Eres un robot? (si o no)', 'no'),
+		array('El fuego es... (caliente o frio)', 'caliente'),
+		array('La estación que viene despues del otoño es...', 'invierno'),
+		array('¿Qué día de la semana es hoy?', strftime('%A')),
+		array('¿En qué mes del año estamos?', strftime('%B')),
+	),
+);
Index: core.php
===================================================================
--- core.php	(revision 3284)
+++ core.php	(working copy)
@@ -3,13 +3,32 @@
 $lang = array
 (
 	'there_can_be_only_one' => 'Solo puede haber una instancia de Kohana por cada página.',
-	'uncaught_exception'    => 'Uncaught %s: %s en el archivo %s en la linea %s',
-	'invalid_method'        => 'Método Inváliudo <tt>%s</tt> llamado en <tt>%s</tt>.',
-	'log_dir_unwritable'    => 'Tu configuración del log.directory no apunta a un directorio con permiso de escritura.',
-	'resource_not_found'    => 'El llamado %s, <tt>%s</tt>, no pudo ser encontrado.',
-	'no_default_route'      => 'Por favor, especificá la ruta en <tt>config/routes.php</tt>.',
-	'no_controller'         => 'Kohana no pudo determinar un controlador para procesar: %s',
-	'page_not_found'        => 'La página que solicitase, <tt>%s</tt>, no pudo ser encontrada.',
-	'stats_footer'          => 'Cargado en {execution_time} segundos, usando {memory_usage} memoria. Generado con Kohana v{kohana_version}.',
-	'error_message'         => 'Ocurrio un error en la linea <strong>line %s</strong> de <strong>%s</strong>.'
+	'uncaught_exception'	=> '%s no capturada: %s en el archivo %s, linea %s',
+	'invalid_method'		=> 'Método inválido %s llamado en %s.',
+	'invalid_property'		=> 'La propiedad %s no existe en la clase %s.',
+	'log_dir_unwritable'	=> 'Tu configuración del &8220;log.directory&8221; no apunta a un directorio con permiso de escritura.',
+	'resource_not_found'	=> 'El archivo de %s con nombre %s, no pudo ser encontrado.',
+	'invalid_filetype'		=> 'El tipo de archivo solicitado, .%s, no esta permitido en la configuración de tus vistas.',
+	'view_set_filename'		=> 'Tienes que definir el nombre de la vista antes de llamar al metodo render',
+	'no_default_route'		=> 'Por favor, especifica la ruta en config/routes.php.',
+	'no_controller'			=> 'Kohana no pudo determinar un controlador para procesar: %s',
+	'page_not_found'		=> 'La página que solicitase, %s, no se encuentra.',
+	'stats_footer'			=> 'Cargado en {execution_time} segundos, usando {memory_usage} de memoria. Generado con Kohana v{kohana_version}.',
+	'error_file_line'		=> '<tt>%s <strong>[%s]:</strong></tt>',
+	'stack_trace'			=> 'Stack Trace',
+	'generic_error'			=> 'Imposible completar la solicitud',
+	'errors_disabled'		=> 'Puedes volver a la <a href="%s">página de inico</a> o <a href="%s">volver a intentarlo</a>.', 
+
+	// Drivers
+	'driver_implements'		=> 'El driver %s para la libreria %s debe implementar el interface %s',
+	'driver_not_found'		=> 'No se ha encontrado el driver %s para la libreria %s',
+
+	// Resource names
+	'config'				=> 'archivo de configuración',
+	'controller'			=> 'controlador',
+	'helper'				=> 'helper',
+	'library'				=> 'librería',
+	'driver'				=> 'driver',
+	'model'					=> 'modelo',
+	'view'					=> 'vista',
 );
\ No newline at end of file
Index: database.php
===================================================================
--- database.php	(revision 3284)
+++ database.php	(working copy)
@@ -3,11 +3,13 @@
 $lang = array
 (
 	'undefined_group'       => 'El grupo %s no esta definido en tu configuración.',
-	'error'                 => 'Hubo un error de SQL: %s',
-	'connection'            => 'Hubo un error conectando a la base de datos: %s',
-	'driver_not_supported'  => 'El driver de base de datos %s no existe.',
+	'error'                 => 'Ocurrió un error de SQL: %s',
+	'connection'            => 'Ocurrió un error conectando a la base de datos: %s',
 	'invalid_dsn'           => 'El DSN que pusiste no es válido: %s',
 	'must_use_set'          => 'Necesitas una clausula SET para tu consulta.',
 	'must_use_where'        => 'Necesitas una clausula WHERE para tu consulta.',
-	'must_use_table'        => 'Necesitas especificar la tabla para tu consulta.'
+	'must_use_table'        => 'Necesitas especificar la tabla para tu consulta.',
+	'table_not_found'       => 'La tabla %s no existe en tu base de datos.',
+	'not_implemented'       => 'El método requerido, %s, no esta soportado por este driver.',
+	'result_read_only'      => 'Los resultados del query son de solo lectura.'
 );
\ No newline at end of file
Index: encrypt.php
===================================================================
--- encrypt.php	(revision 3284)
+++ encrypt.php	(working copy)
@@ -2,6 +2,7 @@
 
 $lang = array
 (
-	'requires_mcrypt'   => 'Para usar la libreria de Encriptación, mcrypt debe estar habilitado.',
-	'no_encryption_key' => 'Para usar la libreria de Encriptación, tenes que especificar una llave de encriptación en tu archivo de configuración.'
+	'undefined_group'   => 'El grupo %s no esta definidp en la configuración.',
+	'requires_mcrypt'   => 'Para usar la librería de Encriptación, mcrypt debe estar habilitado.',
+	'no_encryption_key' => 'Para usar la librería de Encriptación, tienes que especificar una llave de encriptación en tu archivo de configuración.',
 );
Index: errors.php
===================================================================
--- errors.php	(revision 3284)
+++ errors.php	(working copy)
@@ -2,15 +2,15 @@
 
 $lang = array
 (
-	E_KOHANA             => array( 1, 'Error del Framework',   'Revisa la documentación de Kohana para información sobre el siguiente error.'),
-	E_PAGE_NOT_FOUND     => array( 1, 'No se encuentra la página',    'No se encontró la página solicitada. Puede ser que haya sido movida, borrada o archivada.'),
-	E_DATABASE_ERROR     => array( 1, 'Error de Base de Datos',    'Ocurrio un error en la base de datos mientras se ejecutaba el procedimiento requerido. Para más información, revisa el error de abajo.'),
+	E_KOHANA             => array( 1, 'Error del Framework', 'Revisa la documentación de Kohana para información sobre el siguiente error.'),
+	E_PAGE_NOT_FOUND     => array( 1, 'No se encuentra la página', 'No se encontró la página solicitada. Puede ser que haya sido movida, borrada o archivada.'),
+	E_DATABASE_ERROR     => array( 1, 'Error de Base de Datos', 'Ocurrió un error en la base de datos mientras se ejecutaba el procedimiento requerido. Para más información, revisa el error que aparece más abajo.'),
 	E_RECOVERABLE_ERROR  => array( 1, 'Error Recuperable', 'Se detectó un error que evitó que esta página cargara. Si el problema persiste, contacta con el administrador de la web.'),
-	E_ERROR              => array( 1, 'Error Fatal',       ''),
-	E_USER_ERROR         => array( 1, 'Error Fatal',       ''),
-	E_PARSE              => array( 1, 'Error de Syntax',      ''),
-	E_WARNING            => array( 1, 'Advertencia',   ''),
-	E_USER_WARNING       => array( 1, 'Advertencia',   ''),
+	E_ERROR              => array( 1, 'Error Fatal', ''),
+	E_USER_ERROR         => array( 1, 'Error Fatal', ''),
+	E_PARSE              => array( 1, 'Error de Syntaxis', ''),
+	E_WARNING            => array( 1, 'Advertencia', ''),
+	E_USER_WARNING       => array( 1, 'Advertencia', ''),
 	E_STRICT             => array( 2, 'Strict Mode Error', ''),
-	E_NOTICE             => array( 2, 'Runtime Message',   ''),
+	E_NOTICE             => array( 2, 'Runtime Message', ''),
 );
\ No newline at end of file
Index: event.php
===================================================================
--- event.php	(revision 0)
+++ event.php	(revision 0)
@@ -0,0 +1,7 @@
+<?php
+
+$lang = array
+(
+	'invalid_subject' => 'Fallo el intento de añadir el sujeto %s a %s. Los sujetos deben extender la clase Event_Subject.',
+	'invalid_observer' => 'Fallo el intento de añadir el observador %s a %s. Los observadores deben extender la clase Event_Observer.',
+);
Index: image.php
===================================================================
--- image.php	(revision 0)
+++ image.php	(revision 0)
@@ -0,0 +1,27 @@
+<?php 
+
+$lang = array
+(
+	'getimagesize_missing'	  => 'La librería &8220;Image&8221; requiere la función PHP getimagesize, que no parece estar disponible en tu instalación.',
+	'unsupported_method'	  => 'El driver que has elegido en la configuración no soporta el tipo de transformación %s.',
+	'file_not_found'		  => 'La imagen especificada, %s no se ha encontrado. Por favor, verifica que existe utilizando file_exists() antes de manipularla.',
+	'type_not_allowed'		  => 'El tipo de imagen especificado, %s, no es un tipo de imagen permitido.', 
+	'invalid_width'			  => 'El ancho que has especificado, %s, no es valido.',
+	'invalid_height'		  => 'El alto que has especificado, %s, no es valido.',
+	'invalid_dimensions'	  => 'Las dimensiones que has especificado para %s no son validas.',
+	'invalid_master'		  => 'The master dim specified is not valid.',
+	'invalid_flip'			  => 'La dirección de rotación especificada no es valida.',
+	'directory_unwritable'	  => 'El directorio especificado, %s, no tiene permisos de escritura.',
+
+	// ImageMagick specific messages
+	'imagemagick' => array
+	(
+		'not_found' => 'El directorio de ImageMagick especificado, no contiene el programa requrido, %s.', 
+	),
+
+	// GD specific messages
+	'gd' => array
+	(
+		'requires_v2' => 'La librería &8220;Image&8221; requiere GD2. Por favor, lee http://php.net/gd_info para más información.',
+	),
+);
Index: orm.php
===================================================================
--- orm.php	(revision 0)
+++ orm.php	(revision 0)
@@ -0,0 +1,3 @@
+<?php
+
+$lang['query_methods_not_allowed'] = 'No esta permitido utilizar metodos de Query mediante ORM.';
Index: pagination.php
===================================================================
--- pagination.php	(revision 3284)
+++ pagination.php	(working copy)
@@ -2,13 +2,14 @@
 
 $lang = array
 (
+	'undefined_group' => 'El grupo %s no esta definido en la configuracion de la paginación.', 
 	'page'     => 'página',
 	'pages'    => 'páginas',
-	'item'     => 'item',
-	'items'    => 'items',
+	'item'     => 'elemento',
+	'items'    => 'elementos',
 	'of'       => 'de',
 	'first'    => 'primero',
-	'last'     => 'ultimo',
-	'previous' => 'anterior',
-	'next'     => 'siguiente',
+	'last'     => 'último',
+	'previous' => 'Anterior',
+	'next'     => 'Siguiente',
 );
Index: profiler.php
===================================================================
--- profiler.php	(revision 3284)
+++ profiler.php	(working copy)
@@ -5,9 +5,11 @@
 	'benchmarks'   => 'Benchmarks',
 	'post_data'    => 'Datos Posteados',
 	'no_post'      => 'No hay datos posteados',
-	'session_data' => 'Datos de Session',
-	'no_session'   => 'No hay datos de session',
-	'queries'      => 'Consultas a la Base de Datos',
-	'no_queries'   => 'No hay Consultas a la Base de Datos',
-	'no_database'  => 'No se encuentra la Base de Datos',
+	'session_data' => 'Datos de sesión',
+	'no_session'   => 'No hay datos de sesión',
+	'queries'      => 'Consultas a la base de datos',
+	'no_queries'   => 'No hay consultas a la base de datos',
+	'no_database'  => 'No se encuentra la base de datos',
+	'cookie_data'  => 'Datos de la cookie',
+	'no_cookie'    => 'No se encuentran los datos de la cookie',
 );
Index: session.php
===================================================================
--- session.php	(revision 3284)
+++ session.php	(working copy)
@@ -2,6 +2,5 @@
 
 $lang = array
 (
-	'driver_not_supported' => 'El driver de sessiones, %s, no fue encontrado.',
-	'driver_implements'    => 'Los driver de sessiones deben implementar la interfase Session_Driver.'
+	'invalid_session_name' => 'El parametro session_name, %s, no es valido. Solo debe contener caracteres alfanumericos y guiones bajos. También al menos uno debe de ser una letra.',
 );
\ No newline at end of file
Index: swift.php
===================================================================
--- swift.php	(revision 0)
+++ swift.php	(revision 0)
@@ -0,0 +1,6 @@
+<?php
+
+$lang = array
+(
+	'general_error' => 'Ocurrió un error mientras se realizaba el envio del mensaje de correo.',
+);
\ No newline at end of file
Index: upload.php
===================================================================
--- upload.php	(revision 0)
+++ upload.php	(revision 0)
@@ -0,0 +1,6 @@
+<?php
+
+$lang = array
+(
+	'not_writable' => 'El directorio seleccionado, %s, no tiene permisos de escritura.',
+);
\ No newline at end of file
Index: validation.php
===================================================================
--- validation.php	(revision 3284)
+++ validation.php	(working copy)
@@ -3,29 +3,37 @@
 $lang = array
 (
 	// Class errors
-	'error_format'  => 'Tu cadena de mensaje de error, debe contener la cadena {message} .',
 	'invalid_rule'  => 'La regla de validación usada es invalida: %s',
 
 	// General errors
-	'unknown_error' => 'Error de validación desconocido mientras se validaba el %s field.',
+	'unknown_error' => 'Error de validación desconocido al comprobar el campo %s.',
 	'required'      => 'El campo %s es obligatorio.',
-	'min_length'    => 'El campo %s debe tener al menos %d caracteres.',
-	'max_length'    => 'El campo %s debe tener %d caracteres o menos.',
+	'min_length'    => 'El campo %s debe tener un mínimo de %d caracteres.',
+	'max_length'    => 'El campo %s debe tener un máximo de %d caracteres.',
 	'exact_length'  => 'El campo %s debe tener exactamente %d caracteres.',
 	'in_array'      => 'El campo %s debe ser seleccionado de las opciones listadas.',
 	'matches'       => 'El campo %s debe conincidir con el campo %s.',
-	'valid_url'     => 'El campo %s debe contener una url válida, empezando con %s://.',
-	'valid_email'   => 'El campo %s debe contener una dirección de email válida.',
-	'valid_ip'      => 'El campo %s debe contener una direcicón IP válida.',
-	'valid_type'    => 'El campo %s debe contener unicamente %s caracteres.',
+	'valid_url'     => 'El campo %s debe contener una url valida, empezando con %s://.',
+	'valid_email'   => 'El campo %s debe contener una dirección de email valida.',
+	'valid_ip'      => 'El campo %s debe contener una dirección IP valida.',
+	'valid_type'    => 'El campo %s debe contener unicamente %s.',
 	'range'         => 'El campo %s debe estar entre los rangos especificados.',
 	'regex'         => 'El campo %s no coincide con los datos aceptados.',
 	'depends_on'    => 'El campo %s depende del campo %s.',
 
 	// Upload errors
-	'user_aborted'  => 'El archivo %s fue abortado mientras estaba subiendo.',
+	'user_aborted'  => 'El envio del archivo %s fue abortado antes de completarse.',
 	'invalid_type'  => 'El archivo %s no es un tipo de archivo permitido.',
-	'max_size'      => 'El archivo %s que estabas subiendo es muy grande. El tamaño maximo es %s.',
-	'max_width'     => 'El archivo %s debe tener como ancho maximo %s, y tiene %spx.',
-	'max_height'    => 'El archivo %s debe tener como alto maximo %s, y tiene %spx.',
+	'max_size'      => 'El archivo %s que estas enviando es muy grande. El tamaño máximo es de %s.',
+	'max_width'     => 'El archivo %s debe tener como ancho máximo %s, y tiene %spx.',
+	'max_height'    => 'El archivo %s debe tener como alto máximo %s, y tiene %spx.',
+	'min_width'     => 'El archivo %s que estas enviando es muy pequeño. El ancho mínimo permitido es de %spx.',
+	'min_height'    => 'El archivo %s que estas enviando es muy pequeño. El alto mínimo permitido es de %spx.',
+
+	// Field types                                                                                                                                                     
+	'alpha'         => 'caracteres del alfabeto',
+	'alpha_numeric' => 'caracteres del alfabeto y numericos',
+	'alpha_dash'    => 'caracteres del alfabeto, guiones y subrayado',
+	'digit'         => 'digitos',
+	'numeric'       => 'caracteres numéricos',
 );
